tests/constraints/tests.py

```diff
@@ -68,21 +84,99 @@ def test_repr(self):
 
     def test_name(self):
         constraints = get_constraints(Product._meta.db_table)
-        expected_name = 'unique_name'
+        expected_name = 'name_color_uniq'
         self.assertIn(expected_name, constraints)
+
+    def test_constructor_condition_must_be_q(self):
+        with self.assertRaisesMessage(ValueError, 'UniqueConstraint.condition must be a Q instance.'):
+            models.UniqueConstraint(name='uniq', fields=['name'], condition='invalid')
+
+
+@skipIfDBFeature('supports_partial_indexes')
+class ConditionalUniqueConstraintIgnoredTests(TransactionTestCase):
+    available_apps = ['constraints']
+
+    def test_create_conditional_constraint_ignored(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq2',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIsNone(constraint.create_sql(Product, editor))
+            editor.add_constraint(Product, constraint)
+
+    def test_delete_conditional_constraint_ignored(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIsNone(constraint.remove_sql(Product, editor))
+            editor.remove_constraint(Product, constraint)
+
+
+@skipUnlessDBFeature('supports_partial_indexes')
+class ConditionalUniqueConstraintTests(TransactionTestCase):
+    available_apps = ['constraints']
+
+    def test_create_conditional_constraint(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq2',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIn(
+                'WHERE %s IS NULL' % editor.quote_name('color'),
+                str(constraint.create_sql(Product, schema_editor=editor))
+            )
+            editor.add_constraint(Product, constraint)
+        self.assertIn(constraint.name, get_constraints(Product._meta.db_table))
+
+    def test_delete_conditional_constraint(self):
+        constraint = models.UniqueConstraint(
+            name='name_without_color_uniq',
+            fields=['name'],
+            condition=models.Q(color__isnull=True),
+        )
+        with connection.schema_editor() as editor:
+            self.assertIn(constraint.name, get_constraints(Product._meta.db_table))
+            editor.remove_constraint(Product, constraint)
+        self.assertNotIn(constraint.name, get_constraints(Product._meta.db_table))
+
+    def test_database_constraint_with_condition(self):
+        Product.objects.create(name='p1')
+        Product.objects.create(name='p1', color='red')
```

Comment:
Could use `bulk_create`

```suggestion
        Product.objects.bulk_create([
            Product(name='p1'),
            Product(name='p1', color='red'),
        ])
```

Reply:
changed